src/Avalonia.Base/Input/InputManager.cs

```diff
@@ -28,13 +28,52 @@ public class InputManager : IInputManager
         /// <inheritdoc/>
         public IObservable<RawInputEventArgs> PostProcess => _postProcess;
 
+        /// <inheritdoc/>
+        public FocusInputDeviceKind LastInputDeviceType { get; private set; }
+
         /// <inheritdoc/>
         public void ProcessInput(RawInputEventArgs e)
         {
+            SetLastInputDeviceType(e);
+
             _preProcess.OnNext(e);
             e.Device?.ProcessRawEvent(e);
             _process.OnNext(e);
             _postProcess.OnNext(e);
         }
+
+        private void SetLastInputDeviceType(RawInputEventArgs e)
+        {
+            // FocusInputDeviceKind enum
+            // None,            [not valid here]
+            // Mouse,           [supported]
+            // Touch,           [supported*]
+            // Pen,             [not supported yet?]
+            // Keyboard,        [supported]
+            // GameController   [not supported]
+
+            // RawInputEventArgs types
+            //  RawDragEvent -> Ignore, keep existing input type
+            //  RawKeyEventArgs -> Keyboard
+            //  RawPointerEventArgs -> Mouse or ???
+            //   --> RawTouchEventArgs -> Touch
```

Comment:
RawPointerEventArgs can be used for mouse, touch and pen. RawTouchEventArgs can be ignored.

Reply:
But how does `RawPointerEventArgs` differentiate between touch, pen, and mouse? For focus purposes, we really only need keyboard vs. non-keyboard here, but since I made the property public (as it may be useful in other scenarios) it should properly reflect the last input type. 
Though I think proper pen/touch support was still/is still being worked on when I did this so maybe this is improved and there's a better way now?

Reply:
Depending on "Device" property type.

Reply:
Touch support is supported for a while now. Pen support is still on PR blocked by unsolved yet issues.
But even after Pen/WM_POINTER PR merged RawPointerEventArgs.Device is only one place to check input type of RAW event. As for now.